server/src/main/java/org/opensearch/ratelimitting/admissioncontrol/controllers/CPUBasedAdmissionController.java

```diff
@@ -10,46 +10,109 @@
-        this.settings = new CPUBasedAdmissionControllerSettings(clusterSettings, settings);
+    public CPUBasedAdmissionController(
+        String admissionControllerName,
+        ResourceUsageCollectorService resourceUsageCollectorService,
+        ClusterService clusterService,
+        Settings settings
+    ) {
+        super(admissionControllerName, resourceUsageCollectorService, clusterService);
+        this.settings = new CPUBasedAdmissionControllerSettings(clusterService.getClusterSettings(), settings);
     }
 
     /**
-     * This function will take of applying admission controller based on CPU usage
+     * Apply admission control based on process CPU usage
      * @param action is the transport action
      */
     @Override
-    public void apply(String action) {
-        // TODO Will extend this logic further currently just incrementing rejectionCount
+    public void apply(String action, AdmissionControlActionType admissionControlActionType) {
         if (this.isEnabledForTransportLayer(this.settings.getTransportLayerAdmissionControllerMode())) {
-            this.applyForTransportLayer(action);
+            this.applyForTransportLayer(action, admissionControlActionType);
         }
     }
 
-    private void applyForTransportLayer(String actionName) {
-        // currently incrementing counts to evaluate the controller triggering as expected and using in testing so limiting to 10
-        // TODO will update rejection logic further in next PR's
-        if (this.getRejectionCount() < 10) {
-            this.addRejectionCount(1);
+    /**
+     * Apply transport layer admission control if configured limit has been reached
+     */
+    private void applyForTransportLayer(String actionName, AdmissionControlActionType admissionControlActionType) {
+        if (isLimitsBreached(actionName, admissionControlActionType)) {
+            this.addRejectionCount(admissionControlActionType.getType(), 1);
+            if (this.isAdmissionControllerEnforced(this.settings.getTransportLayerAdmissionControllerMode())) {
+                throw new OpenSearchRejectedExecutionException(
+                    String.format("CPU usage admission controller limit reached for action [%s]", admissionControlActionType.name())
+                );
+            }
+        }
+    }
+
+    /**
+     * Check if the configured resource usage limits are breached for the action
+     */
+    private boolean isLimitsBreached(String actionName, AdmissionControlActionType admissionControlActionType) {
+        // check if cluster state is ready
+        if (clusterService.state() != null && clusterService.state().nodes() != null) {
+            long maxCpuLimit = this.getCpuRejectionThreshold(admissionControlActionType);
+            Optional<NodeResourceUsageStats> nodePerformanceStatistics = this.resourceUsageCollectorService.getNodeStatistics(
+                this.clusterService.state().nodes().getLocalNodeId()
```

Comment:
We are always getting here statstics for local node, aren;t we interested here for target data node before sending request there ?

Reply:
Currently we are running this admissionController in the target node. So this will only evaluate local node stats and take the action for the transport calls.